### ISSUE TITLE
[analyzer] InconsistentAnalysisException: Requested result might be inconsistent with previously returned results in analyzer 4.1.0

### ISSUE THREAD

**Issue by noinskit** [dart-lang/build#3312]
After upgrading analyzer from 3.4.1 to 4.1.0 I started getting `InconsistentAnalysisException` exceptions on `flutter pub run build_runner build` such as:
```
2022/05/25 14:03:05 [SEVERE] freezed:freezed on lib/src/parfs/atrial_fibrillation.dart:
2022/05/25 14:03:05 
2022/05/25 14:03:05 InconsistentAnalysisException: Requested result might be inconsistent with previously returned results
2022/05/25 14:03:05 
2022/05/25 14:03:05 package:analyzer/src/dart/analysis/session.dart 133:7  AnalysisSessionImpl._checkConsistency
2022/05/25 14:03:05 package:analyzer/src/dart/analysis/session.dart 94:5   AnalysisSessionImpl.getParsedUnit
2022/05/25 14:03:05 package:build_resolvers/src/resolver.dart 204:32       AnalyzerResolver.compilationUnitFor
2022/05/25 14:03:05 package:build_resolvers/src/resolver.dart 111:26       PerActionResolver.compilationUnitFor.<fn>
2022/05/25 14:03:05 package:source_gen/src/builder.dart 373:18             _hasAnyTopLevelAnnotations
2022/05/25 14:03:05 package:source_gen/src/builder.dart 387:9              _hasAnyTopLevelAnnotations
2022/05/25 14:03:05 package:source_gen/src/builder.dart 89:11              _Builder.build
```

I also opened an issue against freezed first (https://github.com/rrousselGit/freezed/issues/671), but since freezed is not in the stack trace above at all, it might be a problem with the analyzer instead.

It does not happen every time and not always with the same set of files, also tends to differ e.g. between my laptop and my CI/CD env.

I also noticed that it seems related to a change mentioned in analyzer 4.0.0 changelog:
```
Changed synchronous AnalysisSession.getFile and getParsedUnit to fail if there are pending file changes, instead of reading. As for any other AnalysisSession method, await AnalysisContext.applyPendingFileChanges().
```
(`AnalyzerSession.getParsedUnit` is what's throwing the exception here)

I'm using flutter/dart with the following versions:
```
Flutter 3.0.1 • channel stable • https://github.com/flutter/flutter.git
Framework • revision fb57da5f94 (6 days ago) • 2022-05-19 15:50:29 -0700
Engine • revision caaafc5604
Tools • Dart 2.17.1 • DevTools 2.12.2
```

**Comment by jakemac53** [comment]
Actually I realize now this happens not when there is a cycle but when a build step in a later phase resolves a not-yet built asset from a previous phase. But in any case, it is the same issue.

**Comment by jakemac53** [comment]
@noinskit can you try again? (you will need to do another `pub get` I think to get the latest hash from that branch).

**Comment by jakemac53** [comment]
Actually wait nvm I need to make one more change, one second.

**Comment by jakemac53** [comment]
Ok, now it should be ready @noinskit

**Comment by noinskit** [comment]
@jakemac53 I just tested with d4524495fbcdc490c5b1bb001aa96adfdf47f5ae a couple times and everything looks good!

**Comment by jakemac53** [comment]
The fix is published as `build_resolvers` version `2.0.9`, let us know if you see the issue still happening with that version!
